bench/irmin-pack/trace_stat_summary.ml

```diff
@@ -0,0 +1,820 @@
+        let l = List.rev_append c.Def.after.index.new_merge_durations l in
+        l
+    | _ -> l
+  in
+  let finalise = List.rev in
+  Utils.Parallel_folders.folder acc0 accumulate finalise
+
+(** Substract the first and the last timestamps and count the number of ops. *)
+let misc_stats_folder header =
+  let open Def in
+  let acc0 = (0., 0., 0) in
+  let accumulate (t, t', count) = function
+    | `Commit c -> (c.after.timestamp_wall, c.after.timestamp_cpu, count + 1)
+    | _ -> (t, t', count + 1)
+  in
+  let finalise (t, t', count) =
+    ( t -. header.initial_stats.timestamp_wall,
+      t' -. header.initial_stats.timestamp_cpu,
+      count )
+  in
+  Utils.Parallel_folders.folder acc0 accumulate finalise
+
+(* Section 3/4 - Converter from stat_strace to summary *)
+
+(** Fold over [row_seq] and produce the summary.
+
+    {3 Parallel Folders}
+
+    Almost all entries in [t] require to independently fold over the rows of the
+    stat trace, but we want:
+
+    - not to fully load the trace in memory,
+    - not to reread the trace from disk once for each entry,
+    - this current file to be verbose and simple,
+    - to have fun with GADTs and avoid mutability.
+
+    All the boilerplate is hidden behind [Utils.Parallel_folders], a
+    datastructure that holds all folder functions, takes care of feeding the
+    rows to those folders, and preseves the types.
+
+    In the code below, [pf0] is the initial parallel folder, before the first
+    accumulation. Each [|+ ...] statement declares a [acc, accumulate, finalise]
+    triplet, i.e. a folder.
+
+    [val acc : acc] is the initial empty accumulation of a folder.
+
+    [val accumulate : acc -> row -> acc] needs to be folded over all rows of the
+    stat trace. Calling [Parallel_folders.accumulate pf row] will feed [row] to
+    every folders.
+
+    [val finalise : acc -> v] has to be applied on the final [acc] of a folder
+    in order to produce the final value of that folder - which value is meant to
+    be stored in [Trace_stat_summary.t]. Calling [Parallel_folders.finalise pf]
+    will finalise all folders and pass their result to [construct]. *)
+let summarise' header block_count (row_seq : Def.row Seq.t) =
+  let ws = header.Def.word_size / 8 |> float_of_int in
+  let lbs_folder_of_bag_getter value_of_bag =
+    Linear_bag_stat_folder.create header block_count value_of_bag
+  in
+
+  let construct (elapsed_wall, elapsed_cpu, op_count) elapsed_wall_over_blocks
+      elapsed_cpu_over_blocks ops finds cache_misses appended_hashes
+      appended_offsets contents_hash contents_find contents_add node_hash
+      node_mem node_add node_find node_val_v node_val_find node_val_list
+      bytes_read nb_reads bytes_written nb_writes nb_merge merge_durations
+      minor_words promoted_words major_words minor_collections major_collections
+      compactions major_heap_bytes major_heap_top_bytes index_data index_log
+      index_log_async store_dict store_pack watched_nodes =
```

Comment:
We have very few guarantees at the type level as to whether these are in the right order or not, which is somewhat concerning. Could lead to some confusing traces...

Reply:
Good point.

My initial wish was to nest things a bit more (e.g., have a `gc` parameter here instead of `minor_words promoted_words major_words minor_collections major_collections compactions major_heap_bytes major_heap_top_bytes`), but I opted for this easier to code solution.

Now that I look at it again, I might be able to easily setup sub-`parallel folders`. I'll look into that before merging.

Reply:
That went smoothly! https://github.com/mirage/irmin/pull/1404/commits/9f9bd5030be675fb60bd573d1c4457131c9d975d

Reply:
Much nicer, thanks :-)